src/mono/mono/utils/atomic.h

```diff
@@ -266,6 +303,24 @@ static inline gint32 mono_atomic_dec_i32(volatile gint32 *val)
 	return gcc_sync_sub_and_fetch (val, 1);
 }
 
+static inline guint8 mono_atomic_xchg_u8(volatile guint8 *val, guint8 new_val)
+{
+	guint8 old_val;
```

Comment:
For now I wanted to keep the new intrinsics same as the old one.
I assume there might be some historic reasons with old compilers that cause things to be this way, so I don't want to touch it without maintainers asking me to do so.

Reply:
> I am wondering. Why are we not using the `__atomic_exchange_n` bultin from gcc instead of this loop? (This question also applies to the other xchg implementations)

Mostly because mono is old - this code is older than gcc's newer atomics.  The old sync intrinsics didn't have the right operation.

I have a https://github.com/dotnet/runtime/pull/91808 open to will switch us over to C11 atomics on platforms where that makes sense.  But I'm not sure there's a lot of motivation to touch this older implementation as long as GCC doesn't ditch the legacy sync atomics.